File: haskell_study/udemy-learning-path-haskell/230615/05_types_signatures/FormatList.hs
import Data.List


-- we can parenthesize to make more evident how to write those function
--  those are lambda functions which only receive a single parameter
formatList::String->(String->(String->([String]->String)))
-- formatList::String->String->String->[String]->String
formatList start end separator xs = start ++ (intercalate separator (map show xs)) ++ end


main::IO()
main = putStrLn $ formatList "<list>" "</list>" "|" ["first","second","third","fourth"]

File: haskell_study/udemy-learning-path-haskell/230615/05_types_signatures/format_list.py
"""Experiment with lambda expressions in python copying the functionality from
FormatList.hs
"""
# to run this write:
# python3 format_list.py


main = lambda:(
    print(f(2)),\
    print(format_list("<list>")("</list>")("|")(["first","second","third"])),\
    print("hello word"),\
    print(format_list_2("<list>","</list>","|",["first","second","third"]))
)

f = lambda x: x**2

format_list = lambda start: lambda end: lambda separator: lambda xs: start + f"{separator}".join(xs) + end
# format_list(start: str) -> (
#   func(end: str) -> (
#       func(separator)->(
#           func(xs)->str
#       )
#   )
# )

# in lambda we can define multiple arguments for the function
format_list_2 = lambda start, end, separator, xs: start + f"{separator}".join(xs) + end

if __name__ == "__main__":
    main()
